posts/2023-04-13-rust-by-examples.adoc

```diff
@@ -0,0 +1,726 @@
+let double_val = match option_int {
+    None => None,           // all 'arms' of the match must be covered, otherwise compiler error
+    Some(n) => Some(2 * n), // n is the named variable
+};                          // semicolon is mandatory for expression, unless it is a return value
+
+// match statement with enum and value matching:
+match option_int {
+    None => {}              // do-nothing arms has an open-close curly bracket
+    Some(0) => println!("Zero is ignored!"),
+    Some(n) => {            // multi-line expressions or statements are in curly brackets
+        println!("n={}", n);
+        another_side_effect();
+    }                       // no comma needed after curly bracket (but allowed)
+};
+
+// match statement with struct values:
+// discarded values can be represented with _
+match user {
+    User {id, active: true, name: _, roles: _} => println!("User {id} is active!"),
+    User {id, active: false, name: _, roles: _} => println!("User {id} is inactive!"),
+}
+
+// match expression with array
+let array_starting_with_1 = match array {
+    [1, _, _] => Some(array),
+    _ => None,
+};
+
+// match statement with multiple patterns and ranges
+match num {
+    1 | 2 => println!("Small number"),  // multiple pattern
+    3..=7 => println!("Medium number"), // range should be inclusive
+    8..=9 => println!("Almost 10"),
+    ..=0 => println!("Too small!"),     // we may use ..=N or N.. ranges
+    _ => println!("Too big!"),
+}
+----
+
+Pattern matching can also be utilised with ``if let ``, ``while let`` and ``for`` loops:
+[source,rust]
+----
+let bread_spread = Some("butter");
+let mut stack = vec!['a', 'b', 'c'];
+
+// if let:
+if let Some(spread) = bread_spread {
+    println!("The bread has {spread} on it");
+} else {
+    println!("The bread is plain");
+}
+
+// while let:
+while let Some(top) = stack.pop() {
+    println!("{}", top);
+}
+
+// for loop
+for (index, value) in stack.iter().enumerate() {
+    println!("{} is at index {}", value, index);
+}
+----
+
+Pattern matching can also be used with plain ``let`` statements:
+[source,rust]
+----
+// tuple:
+let (a, b) = (2, true);
+
+// this will fail compilation, because the pattern does not match the expression:
+let (a, b, c) = (2, true);
+
+// this will also fail, because the None option is not covered:
+fn foo(opt_value: Option<String>) {
+    let Some(a) = opt_value;
+    ....
+}
+----
+
+Pattern matching can also be used in function parameters:
+[source,rust]
+----
+fn transpose(&(x, y): &(i32, i32)) -> (i32, i32) {
+    (y, x)
+}
+----
+
+Pattern matching tuple function parameters will be useful for closures (invoking lambda expressions).
+
+== Object Oriented Features
+
+Rust implements some OO others not. It is possible to create "objects" to package data and procedures to operate on the object data. String or Vec instances can be considered objects. On the other hand, Rust does not implement inheritance, composition or function overloading.
```

Comment:
* Did you mean _some OO features_?
* Doesn't implement composition? `struct TitleBasic { id: String }` is a composition, maybe I'm missing something here.

Reply:
Silly mistake, of course it is composition.